src/main/java/io/supertokens/webserver/api/bulkimport/AddBulkImportUsers.java

```diff
@@ -0,0 +1,144 @@
+import jakarta.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+
+public class AddBulkImportUsers extends WebserverAPI {
+    private static final int MAX_USERS_TO_ADD = 10000;
+
+    public AddBulkImportUsers(Main main) {
+        super(main, "");
+    }
+
+    @Override
+    public String getPath() {
+        return "/bulk-import/add-users";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        JsonArray users = InputParser.parseArrayOrThrowError(input, "users", false);
+
+        if (users.size() <= 0 || users.size() > MAX_USERS_TO_ADD) {
+            JsonObject errorResponseJson = new JsonObject();
+            String errorMsg = users.size() <= 0 ? "You need to add at least one user."
+                    : "You can only add 10000 users at a time.";
+            errorResponseJson.addProperty("error", errorMsg);
+            throw new ServletException(new WebserverAPI.BadRequestException(errorResponseJson.toString()));
+        }
+
+        AppIdentifier appIdentifier = getTenantIdentifierFromRequest(req).toAppIdentifier();
+
+        JsonArray errorsJson = new JsonArray();
+        ArrayList<BulkImportUser> usersToAdd = new ArrayList<>();
+
+        for (int i = 0; i < users.size(); i++) {
+            try {
+                BulkImportUser user = new BulkImportUser(users.get(i).getAsJsonObject(), null);
+                usersToAdd.add(user);
+
+                for (BulkImportUser.LoginMethod loginMethod : user.loginMethods) {
+                    validateTenantId(appIdentifier, loginMethod.tenantId, loginMethod.recipeId);
+                }
+            } catch (InvalidBulkImportDataException e) {
+                JsonObject errorObj = new JsonObject();
+
+                JsonArray errors = e.errors.stream()
+                        .map(JsonPrimitive::new)
+                        .collect(JsonArray::new, JsonArray::add, JsonArray::addAll);
+
+                errorObj.addProperty("index", i);
+                errorObj.add("errors", errors);
+                errorsJson.add(errorObj);
+            }
+        }
+
+        if (errorsJson.size() > 0) {
+            JsonObject errorResponseJson = new JsonObject();
+            errorResponseJson.addProperty("error",
+                    "Data has missing or invalid fields. Please check the users field for more details.");
+            errorResponseJson.add("users", errorsJson);
+            throw new ServletException(new WebserverAPI.BadRequestException(errorResponseJson.toString()));
+        }
+
+        try {
+            AppIdentifierWithStorage appIdentifierWithStorage = getAppIdentifierWithStorage(req);
```

Comment:
We want to make sure that this API is only called from the public tenant of the app. So we use getAppIdentifierWithStorageFromRequestAndEnforcePublicTenant instead

Reply:
Otherwise devs can call the API with tenant1 (for example), of the app, and then the full json will be added to tenant1's storage, and next time, they can use tenant2, and then that json will go in tenant2's storage. Now most of the time these storages are the same, but, in case they are configured differently, it can cause issues.

Reply:
fixed.